File: Pletor-canvas/backend/app/controllers/workflowy_controller.py
from dataclasses import asdict
from typing import Any

from fastapi import APIRouter, Depends
from sqlalchemy.ext.asyncio import AsyncSession

from app.config.database import get_db
from app.schemas.workflowy import ConfigureWorkFlowyRequest, LinkNodeRequest, UnlinkNodeRequest
from app.services import workflowy_sync_service as wf_service

router = APIRouter(prefix="/api/v1/workflowy")


@router.post("/configure")
async def configure(data: ConfigureWorkFlowyRequest) -> dict[str, str]:
    wf_service.configure_workflowy(data.api_key)
    return {"status": "ok"}


@router.get("/status")
async def status() -> dict[str, bool]:
    return {"configured": wf_service.is_workflowy_configured()}


@router.get("/tree")
async def get_tree() -> list[dict[str, Any]]:
    tree = await wf_service.get_workflowy_tree()
    return [asdict(node) for node in tree]


@router.get("/nodes/{node_id}")
async def get_node(node_id: str) -> dict[str, Any]:
    node = await wf_service.get_workflowy_node(node_id)
    if not node:
        from app.middlewares.error_handler import NotFoundError
        raise NotFoundError("WorkFlowy uzel")
    return asdict(node)


@router.post("/sync/{canvas_id}")
async def sync(canvas_id: str, db: AsyncSession = Depends(get_db)) -> dict[str, int]:
    return await wf_service.sync_workflowy_to_canvas(db, canvas_id)


@router.post("/link")
async def link_node(data: LinkNodeRequest, db: AsyncSession = Depends(get_db)) -> dict[str, str]:
    await wf_service.link_node_to_workflowy(db, data.canvas_node_id, data.workflowy_node_id)
    return {"status": "ok"}


@router.post("/unlink")
async def unlink_node(data: UnlinkNodeRequest, db: AsyncSession = Depends(get_db)) -> dict[str, str]:
    await wf_service.unlink_node_from_workflowy(db, data.canvas_node_id)
    return {"status": "ok"}

File: Pletor-canvas/backend/src/controllers/workflowyController.ts
import type { FastifyRequest, FastifyReply } from 'fastify'
import * as wfSync from '../services/workflowySyncService.js'
import {
  configureWorkflowySchema,
  linkNodeSchema,
  unlinkNodeSchema,
} from '../validators/workflowyValidator.js'

type CanvasIdParam = { canvasId: string }
type NodeIdParam = { nodeId: string }

// Nastav WorkFlowy API klíč
export async function configure(req: FastifyRequest, reply: FastifyReply) {
  const { apiKey } = configureWorkflowySchema.parse(req.body)
  wfSync.configureWorkFlowy(apiKey)
  return reply.send({ status: 'ok', message: 'WorkFlowy připojen' })
}

// Stav připojení
export async function status(_req: FastifyRequest, reply: FastifyReply) {
  return reply.send({ configured: wfSync.isWorkFlowyConfigured() })
}

// Celý WorkFlowy strom (parsovaný)
export async function getTree(_req: FastifyRequest, reply: FastifyReply) {
  const tree = await wfSync.getWorkFlowyTree()
  return reply.send(tree)
}

// Konkrétní WorkFlowy uzel
export async function getNode(req: FastifyRequest<{ Params: NodeIdParam }>, reply: FastifyReply) {
  const node = await wfSync.getWorkFlowyNode(req.params.nodeId)
  if (!node) return reply.status(404).send({ error: 'NOT_FOUND', message: 'WorkFlowy uzel nenalezen' })
  return reply.send(node)
}

// Synchronizace WorkFlowy → Canvas
export async function sync(req: FastifyRequest<{ Params: CanvasIdParam }>, reply: FastifyReply) {
  const result = await wfSync.syncWorkFlowyToCanvas(req.params.canvasId)
  return reply.send(result)
}

// Propojit canvas uzel s WorkFlowy uzlem
export async function linkNode(req: FastifyRequest, reply: FastifyReply) {
  const { canvasNodeId, workflowyNodeId } = linkNodeSchema.parse(req.body)
  const node = await wfSync.linkNodeToWorkFlowy(canvasNodeId, workflowyNodeId)
  return reply.send(node)
}

// Odpojit canvas uzel od WorkFlowy
export async function unlinkNode(req: FastifyRequest, reply: FastifyReply) {
  const { canvasNodeId } = unlinkNodeSchema.parse(req.body)
  const node = await wfSync.unlinkNodeFromWorkFlowy(canvasNodeId)
  return reply.send(node)
}
